internal/torcx/remote.go

```diff
@@ -106,3 +175,416 @@ func parseOsRelease(rd io.Reader) (map[string]string, error) {
+			return nil, errors.Wrapf(err, "failed to load keyrings for %s", name)
+		}
+		var manifest string
+		switch url.Scheme {
+		case "https":
+			tries := 0
+			for {
+				tries++
+				tmpManifest, err := fetchManifest(ctx, url.String())
+				ctxErr := ctx.Err()
+				if err == nil && ctxErr == nil {
+					manifest = tmpManifest
+					break
+				}
+				if ctxErr != nil {
+					return nil, ctxErr
+				}
+				logrus.WithFields(logrus.Fields{
+					"attempt": tries,
+					"name":    name,
+					"error":   err,
+				}).Error("failed to fetch contents manifest")
+				time.Sleep(8 * time.Second)
+			}
+		case "file":
+			path := strings.TrimPrefix(url.String(), "file://")
+			b, err := ioutil.ReadFile(filepath.Clean(path))
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to fetch contents manifest for %s", name)
+			}
+			manifest = string(b)
+		default:
+			return nil, errors.Errorf("unsupported scheme %s", url.Scheme)
+		}
+
+		unwrapped, err := verifyManifest(name, manifest, keyrings)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to verify contents manifest for %s", name)
+		}
+		contents, err := decodeContents(unwrapped)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to decode contents for %s", name)
+		}
+		rc.Contents[name] = *contents
+
+		logrus.WithFields(logrus.Fields{
+			"name": name,
+			"path": path,
+			"url":  url,
+		}).Debug("remote verified")
+	}
+
+	// Length sanity check.
+	if len(rc.Paths) != len(rc.Configs) || len(rc.Paths) != len(rc.Contents) {
+		return nil, errors.Errorf("length mismatch, %d vs %d vs %d", len(rc.Paths), len(rc.Configs), len(rc.Contents))
+	}
+
+	return &rc, nil
+}
+
+// CheckAvailable checks if a given Image is available in the configured remote.
+// On success, it returns the full evaluated base URL for the remote plus and
+// the relative image location.
+func (rc *RemotesCache) CheckAvailable(im Image) (*url.URL, *url.URL, string, error) {
+	if im.Remote == "" {
+		return nil, nil, "", nil
+	}
+	if rc == nil {
+		return nil, nil, "", errors.New("nil RemotesCache")
+	}
+
+	contents, ok := rc.Contents[im.Remote]
+	if !ok {
+		return nil, nil, "", errors.Errorf("manifest for remote %s not found: %s", im.Remote, rc)
+	}
+	config, ok := rc.Configs[im.Remote]
+	if !ok {
+		return nil, nil, "", errors.Errorf("manifest for remote %s not found: %s", im.Remote, rc)
+	}
+	baseURL, err := config.evaluateURL(rc.UsrMountpoint)
+	if err != nil {
+		return nil, nil, "", errors.Wrapf(err, "failed to evaluate URL for %s", im.Remote)
+	}
+	location, hash, err := contents.CheckAvailable(im)
+	if err != nil {
+		return nil, nil, "", errors.Wrapf(err, "inspecting remote %s", im.Remote)
+	}
+	if location == nil {
+		return nil, nil, "", nil
+	}
+
+	return baseURL, location, hash, nil
+}
+
+func verifyManifest(manifestName string, manifest string, keyrings []openpgp.KeyRing) (string, error) {
+	if manifest == "" {
+		return "", errors.New("empty manifest")
+	}
+	if len(keyrings) <= 0 {
```

Comment:
Minor: `== 0`?

Reply:
The builtin `len()` in Golang returns a signed integer (`int`).